internal/vpn/server.go

```diff
@@ -0,0 +1,183 @@
+	if err := conn.Close(); err != nil {
+		s.log.WithError(err).Errorf("Error closing client %s connection", conn.RemoteAddr())
+	}
+}
+
+func (s *Server) serveConn(conn net.Conn) {
+	defer s.closeConn(conn)
+
+	tunIP, tunGateway, err := s.shakeHands(conn)
+	if err != nil {
+		s.log.WithError(err).Errorf("Error negotiating with client %s", conn.RemoteAddr())
+		return
+	}
+
+	tun, err := water.New(water.Config{
+		DeviceType: water.TUN,
+	})
+	if err != nil {
+		s.log.WithError(err).Errorln("Error allocating TUN interface")
+		return
+	}
+	defer func() {
+		s.log.Errorln("DONE SERVING, CLOSING TUN")
+		tunName := tun.Name()
+		if err := tun.Close(); err != nil {
+			s.log.WithError(err).Errorf("Error closing TUN %s", tunName)
+		}
+	}()
+
+	s.log.Infof("Allocated TUN %s", tun.Name())
+
+	if err := SetupTUN(tun.Name(), tunIP.String(), TUNNetmask, tunGateway.String(), TUNMTU); err != nil {
+		s.log.WithError(err).Errorf("Error setting up TUN %s", tun.Name())
+		return
+	}
+
+	connToTunDoneCh := make(chan struct{})
+	tunToConnCh := make(chan struct{})
+	go func() {
+		defer close(connToTunDoneCh)
+
+		if _, err := io.Copy(tun, conn); err != nil {
+			s.log.WithError(err).Errorf("Error resending traffic from VPN client to TUN %s", tun.Name())
+		}
+	}()
+	go func() {
+		defer close(tunToConnCh)
+
+		if _, err := io.Copy(conn, tun); err != nil {
+			s.log.WithError(err).Errorf("Error resending traffic from TUN %s to VPN client", tun.Name())
+		}
+	}()
+
+	// only one side may fail here, so we wait till at least one fails
+	select {
+	case <-connToTunDoneCh:
+	case <-tunToConnCh:
+	}
+}
+
+func (s *Server) shakeHands(conn net.Conn) (tunIP, tunGateway net.IP, err error) {
+	var cHello ClientHello
+	if err := ReadJSON(conn, &cHello); err != nil {
+		return nil, nil, fmt.Errorf("error reading client hello: %w", err)
+	}
+
+	var sHello ServerHello
+
+	for _, ip := range cHello.UnavailablePrivateIPs {
+		if err := s.ipGen.Reserve(ip); err != nil {
+			// this happens only on malformed IP
+			sHello.Status = HandshakeStatusBadRequest
+			if err := WriteJSON(conn, &sHello); err != nil {
+				s.log.WithError(err).Errorln("Error sending server hello")
+			}
+
+			return nil, nil, fmt.Errorf("error reserving IP %s: %w", ip.String(), err)
+		}
+	}
+
+	subnet, err := s.ipGen.Next()
+	if err != nil {
+		sHello.Status = HandshakeNoFreeIPs
+		if err := WriteJSON(conn, &sHello); err != nil {
+			s.log.WithError(err).Errorln("Error sending server hello")
+		}
+
+		return nil, nil, fmt.Errorf("error getting free subnet IP: %w", err)
+	}
+
+	subnetOctets, err := fetchIPv4Octets(subnet)
+	if err != nil {
+		sHello.Status = HandshakeStatusInternalError
+		if err := WriteJSON(conn, &sHello); err != nil {
+			s.log.WithError(err).Errorln("Error sending server hello")
+		}
+
+		return nil, nil, fmt.Errorf("error breaking IP into octets: %w", err)
+	}
+
+	sTUNIP := net.IPv4(subnetOctets[0], subnetOctets[1], subnetOctets[2], subnetOctets[3]+2)
```

Comment:
Could you please add more comments about the logic of `+2`, `+1`, `+4`, `+3` here and below? I think it may increase readability.